File: cli/packages/cli/src/commands/update.ts
import { defineCommand } from "citty";
import {
  loadConfig,
  loadLock,
  saveLock,
  parseSource,
  ensureRepo,
  getRepoPath,
  discoverArtifacts,
  resolveRef,
} from "@agpm/core";

export const updateCommand = defineCommand({
  meta: {
    name: "update",
    description: "Update artifacts to latest versions",
  },
  args: {
    artifact: {
      type: "positional",
      description: "Specific artifact to update (optional, updates all if not specified)",
      required: false,
    },
  },
  async run({ args }) {
    const cwd = process.cwd();
    const config = await loadConfig(cwd);
    const lock = await loadLock(cwd);

    if (config.artifacts.length === 0) {
      console.log("No artifacts configured.");
      return;
    }

    // Filter to specific artifact if provided
    let artifactsToUpdate = config.artifacts;
    if (args.artifact) {
      artifactsToUpdate = config.artifacts.filter(
        (ref) =>
          ref === args.artifact ||
          ref.endsWith(`/${args.artifact}`)
      );

      if (artifactsToUpdate.length === 0) {
        console.log(`Artifact not found: ${args.artifact}`);
        return;
      }
    }

    console.log(`Updating ${artifactsToUpdate.length} artifact(s)...\n`);

    let updated = 0;

    for (const artifactRef of artifactsToUpdate) {
      // Parse artifact reference: source/artifact-name
      const lastSlash = artifactRef.lastIndexOf("/");
      const sourceStr = artifactRef.slice(0, lastSlash);
      const artifactName = artifactRef.slice(lastSlash + 1);

      const parsed = parseSource(sourceStr);

      console.log(`Checking ${artifactRef}...`);

      // Fetch latest
      await ensureRepo(parsed);
      const repoPath = getRepoPath(parsed);

      // Resolve latest SHA
      const latestSha = await resolveRef(repoPath, parsed.ref || "HEAD");

      // Check if update needed
      const lockEntry = lock.artifacts[artifactRef];
      if (lockEntry && lockEntry.sha === latestSha) {
        console.log(`  Already up to date (${latestSha.slice(0, 8)})`);
        continue;
      }

      // Discover to get path
      const artifacts = await discoverArtifacts(repoPath, parsed.subpath);
      const artifact = artifacts.find((a) => a.name === artifactName);

      if (!artifact) {
        console.log(`  Artifact not found in source`);
        continue;
      }

      const oldSha = lockEntry?.sha?.slice(0, 8) || "none";

      // Update lock
      lock.artifacts[artifactRef] = {
        sha: latestSha,
        integrity: `sha256-${latestSha.slice(0, 16)}`,
        path: artifact.path,
        metadata: {
          name: artifact.name,
          description: artifact.description,
        },
      };

      console.log(`  Updated: ${oldSha} -> ${latestSha.slice(0, 8)}`);
      updated++;
    }

    if (updated > 0) {
      await saveLock(cwd, lock);
      console.log(`\nUpdated ${updated} artifact(s) in agpm-lock.json`);
      console.log("Run `agpm install` to apply updates.");
    } else {
      console.log("\nAll artifacts are up to date.");
    }
  },
});

File: src/cli/update.rs
//! Update command - update skills to latest versions

use anyhow::Result;

use crate::config::{SkillsConfig, SkillsLock, LockedSkill};
use crate::git::{GitSource, ClonedRepo, resolve_sha};
use crate::skill::discover_skills;

pub async fn run(skill_name: Option<&str>) -> Result<()> {
    let cwd = std::env::current_dir()?;
    
    let config = SkillsConfig::load(&cwd).await?;
    let mut lock = SkillsLock::load(&cwd).await?;
    
    if config.skills.is_empty() && config.marketplaces.is_empty() {
        println!("No skills configured in skills.json");
        return Ok(());
    }

    let mut updated_count = 0;

    // Update individual skills
    for skill_spec in &config.skills {
        if let Some(name) = skill_name {
            if skill_spec.name != name {
                continue;
            }
        }
        
        println!("Checking {}", skill_spec.name);
        
        let git_source = GitSource::parse(&skill_spec.source)?;
        let new_sha = resolve_sha(&git_source).await?;
        
        let current_sha = lock.skills.get(&skill_spec.name).map(|s| s.sha.as_str());
        
        if current_sha == Some(&new_sha) {
            println!("  Already up to date: {}", &new_sha[..8]);
            continue;
        }
        
        println!("  Updating {} -> {}", 
            current_sha.map(|s| &s[..8]).unwrap_or("none"),
            &new_sha[..8]
        );
        
        // Clone and discover skill
        let cloned = ClonedRepo::clone(&git_source)?;
        let skills = discover_skills(&cloned.path, skill_spec.path.as_deref()).await?;
        
        let skill = skills.into_iter()
            .find(|s| s.metadata.name == skill_spec.name)
            .ok_or_else(|| anyhow::anyhow!("Skill '{}' not found", skill_spec.name))?;
        
        // Update lock
        let locked_skill = LockedSkill {
            name: skill.metadata.name.clone(),
            source: git_source.url.clone(),
            sha: new_sha.clone(),
            path: skill.relative_path.clone(),
            description: Some(skill.metadata.description.clone()),
            marketplace: None,
        };
        lock.skills.insert(skill_spec.name.clone(), locked_skill);
        updated_count += 1;
    }

    // Update marketplace skills
    for marketplace in &config.marketplaces {
        let git_source = GitSource::parse(&marketplace.source)?;
        let new_sha = resolve_sha(&git_source).await?;
        
        for skill_name_to_check in &marketplace.enabled {
            if let Some(name) = skill_name {
                if skill_name_to_check != name {
                    continue;
                }
            }
            
            let current_sha = lock.skills.get(skill_name_to_check).map(|s| s.sha.as_str());
            
            if current_sha == Some(&new_sha) {
                continue;
            }
            
            println!("Updating {} (marketplace: {})", skill_name_to_check, marketplace.name);
            
            let cloned = ClonedRepo::clone(&git_source)?;
            let skills = discover_skills(&cloned.path, None).await?;
            
            let skill = skills.iter()
                .find(|s| &s.metadata.name == skill_name_to_check)
                .ok_or_else(|| anyhow::anyhow!("Skill '{}' not found", skill_name_to_check))?;
            
            let locked_skill = LockedSkill {
                name: skill.metadata.name.clone(),
                source: git_source.url.clone(),
                sha: new_sha.clone(),
                path: skill.relative_path.clone(),
                description: Some(skill.metadata.description.clone()),
                marketplace: Some(marketplace.name.clone()),
            };
            lock.skills.insert(skill_name_to_check.clone(), locked_skill);
            updated_count += 1;
        }
    }

    lock.save(&cwd).await?;
    
    if updated_count > 0 {
        println!("\n{} skill(s) updated in lock file.", updated_count);
        println!("Run 'skills install' to apply updates.");
    } else {
        println!("\nAll skills are up to date.");
    }
    
    Ok(())
}
